Prevent very long compilation runtimes in LateBoundRegionNameCollector

b-naber opened the pull request:
Fixes https://github.com/rust-lang/rust/issues/83150

On recursive types such as in the example given in https://github.com/rust-lang/rust/issues/83150, the current implementation of `LateBoundRegionNameCollector` has very long compilation runtimes. To prevent those we store the types visited in the `middle::ty::Visitor` implementation of `LateBoundRegionNameCollector` in a `SsoHashSet`.

rust-highfive commented:
r? @lcnr

(rust-highfive has picked a reviewer for you, use r? to override)

b-naber commented on src/test/ui/recursion/issue-83150.rs:
```diff
@@ -0,0 +1,9 @@
+fn main() {
```
This test passes for an unexplainable reason. When I compile this file I get the expected error:

```
error[E0275]: overflow evaluating the requirement `Map<&mut Map<&mut Map<&mut Map<&mut Map<.....
```

But the test suite cannot find an error.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/2174825645) [(plain)](https://github.com/rust-lang/rust/commit/22ab0312a0c364c93e6a111593cb22072cdebfc4/checks/2174825645/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.channel         := nightly
configure: rust.debug-assertions := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
tidy error: /checkout/src/test/ui/recursion/issue-83150.rs:7: trailing whitespace
some tidy checks failed


command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/rust-tidy" "/checkout" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "/checkout/obj/build"


failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test --stage 2 src/tools/tidy
Build completed unsuccessfully in 0:00:14
```

</details>

lcnr commented on compiler/rustc_middle/src/ty/print/pretty.rs:
```diff
@@ -1861,18 +1863,47 @@ impl<F: fmt::Write> FmtPrinter<'_, 'tcx, F> {
                 r.super_visit_with(self)
             }
+
+            // In order to prohibit infinite recursion on infinitely recursed types we count the
```
> infinitely recursed types

tnit: there are no infinite types. Types are finite trees, so infinite types are impossible.
We can get very large types though, which causes the hang here.

lcnr reviewed (commented):
I don't think that using type length limit is the best approach here.

The issue here is that the type `T` is `&mut Map<T_prev, closure [T_prev]>` so the type size scales by `2^depth` instead of just `depth`.

Using https://doc.rust-lang.org/nightly/nightly-rustc/rustc_data_structures/sso/struct.SsoHashSet.html in `visit_ty` seems like a potentially better solution to me

b-naber commented:
@lcnr Thanks for the review. 

I don't understand how you would want to use `SsoHashSet`. The sub-`Ty`s we encounter when we traverse these very large types are not the same, so breaking control flow once we have encountered a `Ty` some given number of times is not an option. I don't see any other reason why we would want to collect those types.

lcnr commented:
I don't want to break control at all here. This is not a hang, this is a finite, but incredibly large perf regression.

I want to use a `SsoHashSet` to not look multiple times into identical parts of our type

b-naber commented:
@lcnr Thanks, that clarifies things. You're right, this isn't a hang. 

I updated the PR to use the `SsoHashSet`. There's still problem with the test though. When I changed this to `build-pass` the test fails, but it also fails when I add an expected error to the test with the reason that the test compiled successfully?!

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-10`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/2219434259) [(plain)](https://github.com/rust-lang/rust/commit/6174a2f0a1ba119d23af651cce4f014a84418929/checks/2219434259/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
.................................................................................................... 9400/11720
.................................................................................................... 9500/11720
..............................................................i......i.............................. 9600/11720
.................................................................................................... 9700/11720
........iiiiiii..iiiiii.i........................................................................... 9800/11720
.................................................................................................... 10000/11720
.................................................................................................... 10100/11720
.................................................................................................... 10200/11720
.................................................................................................... 10300/11720
---

---- [ui] ui/anonymous-higher-ranked-lifetime.rs stdout ----
diff of stderr:

114 LL |     h2(|_: (), _: (), _: (), _: ()| {});
115    |     ^^ ---------------------------- found signature of `fn((), (), (), ()) -> _`
116    |     |
-    |     expected signature of `for<'r, 't0> fn(&'r (), Box<(dyn for<'s> Fn(&'s ()) + 'static)>, &'t0 (), for<'s, 't1> fn(&'s (), &'t1 ())) -> _`
+    |     expected signature of `for<'r, 't0> fn(&'r (), Box<(dyn for<'s> Fn(&'s ()) + 'static)>, &'t0 (), for<'s, 't0> fn(&'s (), &'t0 ())) -> _`
118 ...
119 LL | fn h2<F>(_: F) where F: for<'t0> Fn(&(), Box<dyn Fn(&())>, &'t0 (), fn(&(), &())) {}
120    |                         --------------------------------------------------------- required by this bound in `h2`

The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/anonymous-higher-ranked-lifetime/anonymous-higher-ranked-lifetime.stderr
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/anonymous-higher-ranked-lifetime/anonymous-higher-ranked-lifetime.stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args anonymous-higher-ranked-lifetime.rs`
error: 1 errors occurred comparing output.
status: exit status: 1
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/anonymous-higher-ranked-lifetime.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zemit-future-incompat-report" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/anonymous-higher-ranked-lifetime" "-A" "unused" "-Crpath" "-O" "-Cdebuginfo=0" "-Zunstable-options" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/anonymous-higher-ranked-lifetime/auxiliary"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
error[E0631]: type mismatch in closure arguments
  --> /checkout/src/test/ui/anonymous-higher-ranked-lifetime.rs:2:5
   |
LL |     f1(|_: (), _: ()| {}); //~ ERROR type mismatch
   |     ^^ -------------- found signature of `fn((), ()) -> _`
   |     |
   |     expected signature of `for<'r, 's> fn(&'r (), &'s ()) -> _`
...
LL | fn f1<F>(_: F) where F: Fn(&(), &()) {}
   |                         ------------ required by this bound in `f1`
error[E0631]: type mismatch in closure arguments
  --> /checkout/src/test/ui/anonymous-higher-ranked-lifetime.rs:3:5
   |
   |
LL |     f2(|_: (), _: ()| {}); //~ ERROR type mismatch
   |     ^^ -------------- found signature of `fn((), ()) -> _`
   |     |
   |     expected signature of `for<'a, 'r> fn(&'a (), &'r ()) -> _`
...
LL | fn f2<F>(_: F) where F: for<'a> Fn(&'a (), &()) {}
   |                         ----------------------- required by this bound in `f2`
error[E0631]: type mismatch in closure arguments
  --> /checkout/src/test/ui/anonymous-higher-ranked-lifetime.rs:4:5
   |
   |
LL |     f3(|_: (), _: ()| {}); //~ ERROR type mismatch
   |     ^^ -------------- found signature of `fn((), ()) -> _`
   |     |
   |     expected signature of `for<'r> fn(&(), &'r ()) -> _`
...
LL | fn f3<'a, F>(_: F) where F: Fn(&'a (), &()) {}
   |                             --------------- required by this bound in `f3`
error[E0631]: type mismatch in closure arguments
  --> /checkout/src/test/ui/anonymous-higher-ranked-lifetime.rs:5:5
   |
   |
LL |     f4(|_: (), _: ()| {}); //~ ERROR type mismatch
   |     ^^ -------------- found signature of `fn((), ()) -> _`
   |     |
   |     expected signature of `for<'s, 'r> fn(&'s (), &'r ()) -> _`
...
LL | fn f4<F>(_: F) where F: for<'r> Fn(&(), &'r ()) {}
   |                         ----------------------- required by this bound in `f4`
error[E0631]: type mismatch in closure arguments
  --> /checkout/src/test/ui/anonymous-higher-ranked-lifetime.rs:6:5
   |
   |
LL |     f5(|_: (), _: ()| {}); //~ ERROR type mismatch
   |     ^^ -------------- found signature of `fn((), ()) -> _`
   |     |
   |     expected signature of `for<'r> fn(&'r (), &'r ()) -> _`
...
LL | fn f5<F>(_: F) where F: for<'r> Fn(&'r (), &'r ()) {}
   |                         -------------------------- required by this bound in `f5`
error[E0631]: type mismatch in closure arguments
  --> /checkout/src/test/ui/anonymous-higher-ranked-lifetime.rs:7:5
   |
   |
LL |     g1(|_: (), _: ()| {}); //~ ERROR type mismatch
   |     ^^ -------------- found signature of `fn((), ()) -> _`
   |     |
   |     expected signature of `for<'r> fn(&'r (), Box<(dyn for<'s> Fn(&'s ()) + 'static)>) -> _`
...
LL | fn g1<F>(_: F) where F: Fn(&(), Box<dyn Fn(&())>) {}
   |                         ------------------------- required by this bound in `g1`
error[E0631]: type mismatch in closure arguments
  --> /checkout/src/test/ui/anonymous-higher-ranked-lifetime.rs:8:5
   |
   |
LL |     g2(|_: (), _: ()| {}); //~ ERROR type mismatch
   |     ^^ -------------- found signature of `fn((), ()) -> _`
   |     |
   |     expected signature of `for<'r> fn(&'r (), for<'s> fn(&'s ())) -> _`
...
LL | fn g2<F>(_: F) where F: Fn(&(), fn(&())) {}
   |                         ---------------- required by this bound in `g2`
error[E0631]: type mismatch in closure arguments
  --> /checkout/src/test/ui/anonymous-higher-ranked-lifetime.rs:9:5
   |
   |
LL |     g3(|_: (), _: ()| {}); //~ ERROR type mismatch
   |     ^^ -------------- found signature of `fn((), ()) -> _`
   |     |
   |     expected signature of `for<'s> fn(&'s (), Box<(dyn for<'r> Fn(&'r ()) + 'static)>) -> _`
...
LL | fn g3<F>(_: F) where F: for<'s> Fn(&'s (), Box<dyn Fn(&())>) {}
   |                         ------------------------------------ required by this bound in `g3`
error[E0631]: type mismatch in closure arguments
  --> /checkout/src/test/ui/anonymous-higher-ranked-lifetime.rs:10:5
   |
   |
LL |     g4(|_: (), _: ()| {}); //~ ERROR type mismatch
   |     ^^ -------------- found signature of `fn((), ()) -> _`
   |     |
   |     expected signature of `for<'s> fn(&'s (), for<'r> fn(&'r ())) -> _`
...
LL | fn g4<F>(_: F) where F: Fn(&(), for<'r> fn(&'r ())) {}
   |                         --------------------------- required by this bound in `g4`
error[E0631]: type mismatch in closure arguments
  --> /checkout/src/test/ui/anonymous-higher-ranked-lifetime.rs:11:5
   |
   |
LL |     h1(|_: (), _: (), _: (), _: ()| {}); //~ ERROR type mismatch
   |     ^^ ---------------------------- found signature of `fn((), (), (), ()) -> _`
   |     |
   |     expected signature of `for<'r, 's> fn(&'r (), Box<(dyn for<'t0> Fn(&'t0 ()) + 'static)>, &'s (), for<'t0, 't1> fn(&'t0 (), &'t1 ())) -> _`
...
LL | fn h1<F>(_: F) where F: Fn(&(), Box<dyn Fn(&())>, &(), fn(&(), &())) {}
   |                         -------------------------------------------- required by this bound in `h1`
error[E0631]: type mismatch in closure arguments
  --> /checkout/src/test/ui/anonymous-higher-ranked-lifetime.rs:12:5
   |
   |
LL |     h2(|_: (), _: (), _: (), _: ()| {}); //~ ERROR type mismatch
   |     ^^ ---------------------------- found signature of `fn((), (), (), ()) -> _`
   |     |
   |     expected signature of `for<'r, 't0> fn(&'r (), Box<(dyn for<'s> Fn(&'s ()) + 'static)>, &'t0 (), for<'s, 't0> fn(&'s (), &'t0 ())) -> _`
...
LL | fn h2<F>(_: F) where F: for<'t0> Fn(&(), Box<dyn Fn(&())>, &'t0 (), fn(&(), &())) {}
   |                         --------------------------------------------------------- required by this bound in `h2`
error: aborting due to 11 previous errors

For more information about this error, try `rustc --explain E0631`.


------------------------------------------


---- [ui] ui/recursion/issue-83150.rs stdout ----

error: ui test compiled successfully!
status: exit status: 0
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/recursion/issue-83150.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zemit-future-incompat-report" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/recursion/issue-83150" "-A" "unused" "-Crpath" "-O" "-Cdebuginfo=0" "-Zunstable-options" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/recursion/issue-83150/auxiliary"
------------------------------------------

------------------------------------------
stderr:
---

Some tests failed in compiletest suite=ui mode=ui host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu


command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/checkout/src/test/ui" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--suite" "ui" "--mode" "ui" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-10/bin/FileCheck" "--nodejs" "/usr/bin/node" "--host-rustcflags" "-Crpath -O -Cdebuginfo=0 -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Cdebuginfo=0 -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python3" "--lldb-python" "/usr/bin/python3" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "10.0.0" "--llvm-components" "aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets amdgpu amdgpuasmparser amdgpucodegen amdgpudesc amdgpudisassembler amdgpuinfo amdgpuutils analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo cfguard codegen core coroutines coverage debuginfocodeview debuginfodwarf debuginfogsym debuginfomsf debuginfopdb demangle dlltooldriver dwarflinker engine executionengine frontendopenmp fuzzmutate globalisel hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo instcombine instrumentation interpreter ipo irreader jitlink lanai lanaiasmparser lanaicodegen lanaidesc lanaidisassembler lanaiinfo libdriver lineeditor linker lto mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts object objectyaml option orcerror orcjit passes perfjitevents powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo riscvutils runtimedyld scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target textapi transformutils vectorize webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info x86utils xcore xcorecodegen xcoredesc xcoredisassembler xcoreinfo xray" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--color" "always"


failed to run: /checkout/obj/build/bootstrap/debug/bootstrap --stage 2 test --exclude src/tools/tidy
Build completed unsuccessfully in 0:14:05
```

</details>

b-naber commented:
There's also a problem with another test, seems as if `LateBoundRegionNameCollector' has different functionality if we skip types we've seen before.

lcnr commented on compiler/rustc_middle/src/ty/print/pretty.rs:
```diff
@@ -1861,18 +1863,38 @@ impl<F: fmt::Write> FmtPrinter<'_, 'tcx, F> {
+            fn visit_ty(&mut self, ty: Ty<'tcx>) -> ControlFlow<Self::BreakTy> {
+                debug!("LateBoundRegionNameCollector::visit_ty(ty: {:?}", ty);
+                let not_previously_inserted = self.type_collector.insert(ty);
+                if not_previously_inserted { ty.super_visit_with(self) } else { ControlFlow::BREAK }
```
```suggestion
                if not_previously_inserted { ty.super_visit_with(self) } else { ControlFlow::CONTINUE }
```
you must not break here, otherwise you don't see `T2` in `Type<T1, T1, T2>`

b-naber commented:
@lcnr Can you take another look? What do we do about the tests?

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-10`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/2221584890) [(plain)](https://github.com/rust-lang/rust/commit/08a8b83a5a0b2daf8a3dbc04982d595965069a3b/checks/2221584890/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
.................................................................................................... 9400/11720
.................................................................................................... 9500/11720
..............................................................i......i.............................. 9600/11720
.................................................................................................... 9700/11720
........iiiiiii...iiiiiii........................................................................... 9800/11720
.................................................................................................... 10000/11720
.................................................................................................... 10100/11720
.................................................................................................... 10200/11720
.................................................................................................... 10300/11720
---
---- [ui] ui/recursion/issue-83150.rs stdout ----

error: ui test compiled successfully!
status: exit status: 0
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/recursion/issue-83150.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zemit-future-incompat-report" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/recursion/issue-83150" "-A" "unused" "-Crpath" "-O" "-Cdebuginfo=0" "-Zunstable-options" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/recursion/issue-83150/auxiliary"
------------------------------------------

------------------------------------------
stderr:
---
test result: FAILED. 11623 passed; 1 failed; 96 ignored; 0 measured; 0 filtered out; finished in 144.05s



command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/checkout/src/test/ui" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--suite" "ui" "--mode" "ui" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-10/bin/FileCheck" "--nodejs" "/usr/bin/node" "--host-rustcflags" "-Crpath -O -Cdebuginfo=0 -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Cdebuginfo=0 -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python3" "--lldb-python" "/usr/bin/python3" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "10.0.0" "--llvm-components" "aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets amdgpu amdgpuasmparser amdgpucodegen amdgpudesc amdgpudisassembler amdgpuinfo amdgpuutils analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo cfguard codegen core coroutines coverage debuginfocodeview debuginfodwarf debuginfogsym debuginfomsf debuginfopdb demangle dlltooldriver dwarflinker engine executionengine frontendopenmp fuzzmutate globalisel hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo instcombine instrumentation interpreter ipo irreader jitlink lanai lanaiasmparser lanaicodegen lanaidesc lanaidisassembler lanaiinfo libdriver lineeditor linker lto mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts object objectyaml option orcerror orcjit passes perfjitevents powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo riscvutils runtimedyld scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target textapi transformutils vectorize webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info x86utils xcore xcorecodegen xcoredesc xcoredisassembler xcoreinfo xray" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--color" "always"


failed to run: /checkout/obj/build/bootstrap/debug/bootstrap --stage 2 test --exclude src/tools/tidy
Build completed unsuccessfully in 0:14:15
```

</details>

lcnr commented:
i think the correct attribute here is `// build-fail`. By default we only try to check (like `cargo check`) ui tests afaik.

b-naber commented:
@lcnr Yes, that's the correct attribute. Thanks for the review and the help.

lcnr commented on compiler/rustc_middle/src/ty/print/pretty.rs:
```diff
@@ -1861,18 +1863,42 @@ impl<F: fmt::Write> FmtPrinter<'_, 'tcx, F> {
-        let mut collector = LateBoundRegionNameCollector(&mut self.used_region_names);
+        let mut collector = LateBoundRegionNameCollector {
+            used_region_names: &mut self.used_region_names,
+            type_collector: SsoHashSet::with_capacity(1000),
```
```suggestion
            type_collector: SsoHashSet::new(),
```
we don't want to allocate here by default.

lcnr commented on src/test/ui/recursion/issue-83150.rs:
```diff
@@ -0,0 +1,11 @@
+// build-fail
+  //~^ overflow evaluating
```
```suggestion
//~^ overflow evaluating
```

lcnr commented:
update the PR name and description, then this should be ready after a perf run.

b-naber commented:
@lcnr This is ready for the perf run.

bors commented:
:umbrella: The latest upstream changes (presumably #76814) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

lcnr commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit 3194b26ab00447662897fdfb98c01160456f9e08 with merge 9fb1f198bce56ca989b0203c337bedffb89b96a1...
<!-- homu: {"type":"TryBuildStarted","head_sha":"3194b26ab00447662897fdfb98c01160456f9e08","merge_sha":"9fb1f198bce56ca989b0203c337bedffb89b96a1"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/2268671965)
Build commit: 9fb1f198bce56ca989b0203c337bedffb89b96a1 (`9fb1f198bce56ca989b0203c337bedffb89b96a1`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/2268671965"},"merge_sha":"9fb1f198bce56ca989b0203c337bedffb89b96a1"} -->

rust-timer commented:
Queued 9fb1f198bce56ca989b0203c337bedffb89b96a1 with parent b1b0a1597c40b55b81e5b95ea5fc825a1287b2bb, future [comparison URL](https://perf.rust-lang.org/compare.html?start=b1b0a1597c40b55b81e5b95ea5fc825a1287b2bb&end=9fb1f198bce56ca989b0203c337bedffb89b96a1).

rust-timer commented:
Finished benchmarking try commit (9fb1f198bce56ca989b0203c337bedffb89b96a1): [comparison url](https://perf.rust-lang.org/compare.html?start=b1b0a1597c40b55b81e5b95ea5fc825a1287b2bb&end=9fb1f198bce56ca989b0203c337bedffb89b96a1).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. Please note that if the perf results are neutral, you should likely undo the rollup=never given below by specifying `rollup-` to bors.

Importantly, though, if the results of this run are non-neutral **do not** roll this PR up -- it will mask other regressions or improvements in the roll up.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf

lcnr commented:
@bors r+

bors commented:
:pushpin: Commit 3194b26ab00447662897fdfb98c01160456f9e08 has been approved by `lcnr`

<!-- @bors r=lcnr 3194b26ab00447662897fdfb98c01160456f9e08 -->
<!-- homu: {"type":"Approved","sha":"3194b26ab00447662897fdfb98c01160456f9e08","approver":"lcnr"} -->

bors commented:
:hourglass: Testing commit 3194b26ab00447662897fdfb98c01160456f9e08 with merge d203fceeb12f8c0e3123fc45036575018d2f990a...
<!-- homu: {"type":"BuildStarted","head_sha":"3194b26ab00447662897fdfb98c01160456f9e08","merge_sha":"d203fceeb12f8c0e3123fc45036575018d2f990a"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/2273063534)
Approved by: lcnr
Pushing d203fceeb12f8c0e3123fc45036575018d2f990a to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"lcnr","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/2273063534"},"merge_sha":"d203fceeb12f8c0e3123fc45036575018d2f990a"} -->

estebank commented:
Should this be backported to 1.52? @rust-lang/compiler
